### ISSUE TITLE
Cannot connect to foreign Oracle server

### ISSUE THREAD

**Issue by JimOL** [laurenz/oracle_fdw#71]
I can connect to my oracle database with sqlplus with the following string:

sqlplus hazus_db/mypassword@host1:1521/gis1.vancouver.ca

I make the foreign server like this:

create server hazus_db foreign data wrapper oracle_fdw
OPTIONS(dbserver 'hazus_db/mypassword@host1:1521/gisdev.vancouver.ca');

I make the user mapping like this:

create user mapping for postgres server hazus_db
options (user 'HAZUS_DB', password 'mypassword')

Next I try:
SELECT oracle_diag('hazus_db'); 

The result is the message:

ERROR:  cannot connect to foreign Oracle server
DETAIL:  ORA-12154: TNS:could not resolve the connect identifier specified

The postgresql log says:

2016-03-18 15:08:47 PDT ERROR:  cannot connect to foreign Oracle server
2016-03-18 15:08:47 PDT DETAIL:  ORA-12154: TNS:could not resolve the connect identifier specified

I am using PostgreSQL 9.2.
sqlplus tells me that my Oracle version  is Oracle Database 11g Enterprise Edition Release 11.2.0.4.0 64 bit

Any suggestions?

Thanks

**Comment by laurenz** [comment]
Yes, I think I know what is going wrong.

First, is <tt>gis1.vancouver.ca</tt> the same as <tt>gisdev.vancouver.ca</tt>?
That's one difference that I spot right away.

Second, the "dbserver" option should not include user/password and the at-sign.
So it should be:

    CREATE SERVER hazus_db FOREIGN DATA WRAPPER oracle_fdw
       OPTIONS(dbserver 'host1:1521/gisdev.vancouver.ca');

**Comment by JimOL** [comment]
That did it, Laurenz. I am connected to Oracle thru PostgreSQL. Woo hoo!

Thanks for your great work on this extension.
Jim
Jim O'Leary
Programmer Analyst
GIS and CADD Services Branch
City of Vancouver
Tel: 604.871.6295
Fax: 604.871.6119
3rd Floor – 814 Richards St, Vancouver  BC V6B 3A7

**Comment by JimOL** [comment]
Hi Laurenz

I have now created a foreign table from Oracle Spatial, but when I try to show it in QGIS, it says, “invalid layer – cannot load”. I think the geometry of the foreign table is correct, because when I query in pgAdmin:

Select st_astext(geom)
From public.streets_hazus

…it shows me what appears to be a correct WKT, that is:

LINESTRING(-123.12345,49.12345, -123.678, 49.678)

It would be very useful for us to show  our Oracle tables via Postgres/PostGIS layers in QGIS. Is it possible? Or is this something that I should post to the QGIS developers?

Thanks
Jim
Jim O'Leary
Programmer Analyst
GIS and CADD Services Branch
City of Vancouver
Tel: 604.871.6295
Fax: 604.871.6119
3rd Floor – 814 Richards St, Vancouver  BC V6B 3A7

**Comment by laurenz** [comment]
I have never heard of QGIS, so I cannot help there.
It might be useful to find out what that error message means; maybe asking QGIS could shed some light on this.

@vmora, do you have an idea?

**Comment by JimOL** [comment]
Hi Laurenz

I think the problem is that the foreign table does not have a spatial index on it or a primary key that QGIS recognizes. I created a “local” table from the foreign table in Postgres by:

create table public. streets_hazus_local as
select * from streets_hazus

Then I added a spatial index to the table:

CREATE INDEX streets_hazus_local_geom_idx
ON streets_hazus_local
USING gist
(geom);

Then I added the auto increment field:

ALTER TABLE streets_hazus_local ADD COLUMN id SERIAL;
UPDATE streets_hazus_local SET id = DEFAULT;
ALTER TABLE streets_hazus_local ADD PRIMARY KEY (id);

After this I could see the “local” table as well as the foreign table in the QGIS interface:


[cid:image002.png@01D18B43.AC7A2210]
…and QGIS renders the streets_hazus_local table onto the map. That tells me that the geometry in the foreign table is correct, but QGIS needs either the spatial index or the primary key to render it on the map. I’ll take up the case with the QGIS folks.

Thanks again for your great extension.
Jim
Jim O'Leary
Programmer Analyst
GIS and CADD Services Branch
City of Vancouver
Tel: 604.871.6295
Fax: 604.871.6119
3rd Floor – 814 Richards St, Vancouver  BC V6B 3A7
